### DISCUSSION TITLE
How to Save image locally - Form Component

### DISCUSSION THREAD

**Discussion by Pieter3033** [sqlpage/SQLPage#150] [Q&A] [answered]
Hi,

I am struggling to save a image locally.
OS: Windows

**Configuration File:**
`"allow_exec" : true`

**Code:**
```sql
select 'form' as component, 'Upload a new image' as title, 'upload.sql' as action;
select 'text' as type, 'Title' as name, true as required;
select 'text' as type, 'Description' as name;
select 'file' as type, 'Image' as name, 'image/*' as accept;
```

```sql
set file_path = sqlpage.uploaded_file_path('Image');
select sqlpage.exec('mv', $file_path , 'C:\New');
```

**Error:**
![image](https://github.com/lovasoa/SQLpage/assets/141907966/25b02837-c5da-40e0-a24e-69e68964cf19)

**Comment by lovasoa** [comment]
It looks like you are using windows. The `mv` command is specific to posix systems (MacOS and Linux). The equivalent on windows is `move`. I'll update the documentation to specify that.

https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/move
https://stackoverflow.com/questions/17607612/equivalent-of-rm-and-mv-in-windows-cmd

**Reply by lovasoa** [reply]
I updated the documentation: https://sql.ophir.dev/functions.sql?function=uploaded_file_path#function

**Reply by Pieter3033** [reply]
Hi @lovasoa, Thank you for the feedback.

I have updated the code accordingly, but it still generates the same results:
![image](https://github.com/lovasoa/SQLpage/assets/141907966/e6046573-1872-4ecf-8184-2c45de4ad0f8)


I attempted to run the commands in cmd afterwards and it generated this result:
![image](https://github.com/lovasoa/SQLpage/assets/141907966/452c4675-ddf4-4d09-a74b-6089f8563ed3)

I then checked the temp file as I attempted to upload a image, it does not seem to display in the temp file? 
I could be wrong but I have examined the temp file thoroughly.


I then used the [updated ](https://sql.ophir.dev/functions.sql?function=uploaded_file_path#function) code:

```
SET file_name = sqlpage.random_string(10);
SET images = sqlpage.uploaded_file_path('Image');
SELECT sqlpage.exec('move', sqlpage.uploaded_file_path('Image') , 'C:\New' || $file_name);
```

Error generated:

`exec('', sqlpage.uploaded_file_path(''), 'C:\New' || $file_name) is not a valid call. Only variables (such as $my_variable) and sqlpage function calls (such as sqlpage.header('my_header')) are supported as arguments to sqlpage functions.`

**Reply by Pieter3033** [reply]
**Note:** using Microsoft SQL Server.

I have attempted to store the image directly inside of my database, exactly as you did in the [image gallery with user uploads](https://github.com/lovasoa/SQLpage/tree/main/examples/image%20gallery%20with%20user%20uploads) example.
The value keeps writing in as NULL.

I have tried the following datatypes in the database:
- nvarchar(max)
- varchar(max)
- image
- text

The reason I want to store it locally is we are going to have to use images for logging purposes and I would like to push the images to a image to text api as well as a barcode scanner api to prevent human error.

**Reply by lovasoa** [reply, marked as answer]
> I have updated the code accordingly, but it still generates the same results

Sorry, I'm not familiar with windows. I think `move` is a built-in in cmd.exe and cannot be executed as an external program. Can you try  `sqlpage.exec('cmd', '/C', 'move', $file_path , 'C:\New')` ?

> `SELECT sqlpage.exec('move', sqlpage.uploaded_file_path('Image') , 'C:\New' || $file_name);`

As stated in the error message, this is not supported. You need to do the concatenation separately and use just a variable name as a function parameter:

```sql
SELECT sqlpage.some_function('a' || 'b'); -- INCORRECT


SET $result = 'a' || 'b'
SELECT sqlpage.some_function($result); -- CORRECT
```

> the value keeps writing in as NULL.

All sqlpage parameters are bound to your queries as `VARCHAR(MAX)`. Can you please provide a small self-contained reproduction ?
If there is indeed a bug, I'll need to be able to reproduce it locally. A tiny sqlpage website that works out-of the box on an empty sql server database would be ideal.

**Reply by Pieter3033** [reply]
> sqlpage.exec('cmd', '/C', 'move', $file_path , 'C:\New')

**Code:**
```sql
SET images = sqlpage.uploaded_file_path('my_file');
SELECT sqlpage.exec('Move-Item' , $images , 'C:\New' );
```
**Error:**
`Command 'cmd' failed with exit code exit code: 1: The system cannot find the file specified.`


**Test:**
I just tested it with a normal text file and it **worked** 100%:
```sql
SET images = 'C:\test\test.txt';
SELECT sqlpage.exec('cmd', '/C', 'move', $images, 'C:\New');
```

> I then checked the temp file as I attempted to upload a image, it does not seem to display in the temp file?

Is there a way to locate the image within the temp file?

**Reply by Pieter3033** [reply]
```
All sqlpage parameters are bound to your queries as VARCHAR(MAX). Can you please provide a small self-contained reproduction ?
If there is indeed a bug, I'll need to be able to reproduce it locally. A tiny sqlpage website that works out-of the box on an empty sql server database would be ideal.
```
Tested it in a small self contained reproduction. It's working now. Not sure what I did wrong previously.

Still struggling with saving to local directory.

**Reply by Pieter3033** [reply]
Hi @lovasoa,

Thank you for your time and assistance.

I figured out what the problem is and it was on my side.
I customized my form component in the previous version of SQLPage and renamed it to form_new.handlebars and never updated it accordingly after SQLPage 0.17.0.

I apologize for wasting your time on this one.

It is saving now to the correct directory with the code:
```sql
SET images = sqlpage.uploaded_file_path('Image');
SELECT sqlpage.exec('cmd', '/C', 'move', $images, 'C:\New');
```

**Reply by lovasoa** [reply]
Ok, great ! For the future, maybe you should name your custom components differently, so you can still try the vanilla components when they are updated !
